### PR TITLE
feat(a11y): xy cartesian data table

### PR DESCRIPTION
**Pull request by rshen91**
## Summary
Cartesian charts now have a data table to provide alternative text descriptions of the chart canvas data. A caption was added to the table to provide more context for users.

For charts where there are more than 20 data points, the table will be similar to the legend. Users can compare the series at the same x interval and click to see more data or previous data. 

For small data sets, users will be able to access the full data set. For charts with small datasets but multiple series, the x values of the points are grouped to provide more a greater idea of comparison between the series. 

- [x] Get the data formatted appropriately for the component (have the full data set formatted vs. how legend items are truncated at last values)
- [x] Fix the show next data for the table
- [x] Small multiple charts
- [x] x Value formatting
- [x] Clean up the caption and any extraneous data for the xy table component (axes or not etc.)
- [x] Clean up keys in the component (no Math.random())
- [x] Limit the comparison by x value behavior for one series or data points that are less than 20 (double check that limit with the current one used in partition charts) 2d399b6
- [x] Clarify the small multiple category specification: the small multiple categories used to describe the vertical and horizontal position of the small multiple needs to be described in the table as columns 8f88313
- [x] Confirm navigation with screen reader/Voice Over - met with Zuhair and Trevor from EUI about usability 
- [x] Add titles from axes to the caption for greater semantic context for users f2d9a35
- [x] Confirmed table is read in Kibana when locally testing
- [x] Write tests
- [x] Add `debugA11y` for developer debugging of the screen reader table

## Details

Unlike the partitions specific data table, the cartesian data table will show a value per series for easier comparison.

- Future refactoring: The current code uses data from Last Values logic: can we find a way to reduce code duplication? Reducing duplication can avoid losing a fix that was deployed only on a particular part of the code and not reported here

## Issues
Part of https://github.com/elastic/elastic-charts/issues/300

### Checklist

- [x] The proper **chart type** label has been added (e.g. `:xy`, `:partition`)
- [x] The proper **feature** labels have been added (e.g. `:interactions`, `:axis`)
- [x] All related issues have been linked (i.e. `closes #123`, `fixes #123`)
- [x] Unit tests have been added or updated to match the most common scenarios
- [x] The proper documentation and/or storybook story has been added or updated
- [x] The code has been checked for cross-browser compatibility (Chrome, Firefox, Safari, Edge)
- [x] Visual changes have been tested with all available themes including `dark`, `light`, `eui-dark` & `eui-light`

### CODE REVIEW

**Comment by rshen91** [comment]
Jenkins test this please

File: packages/charts/src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx [hunk 1]
```diff
@@ -162,25 +163,25 @@ class XYChartComponent extends React.Component<XYChartProps> {
...
+            width,
+            height,
+          }}
+          // eslint-disable-next-line jsx-a11y/no-interactive-element-to-noninteractive-role
+          role="presentation"
+        >
+          <ScreenReaderSummary />
+          {!debug && <ScreenReaderCartesianTable />}
+        </canvas>
+        {debug && <ScreenReaderCartesianTable />}
```
Comment (nickofthyme) [inline packages/charts/src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx:183]: Debug is only used internally. Is this table visibility meant to be toggle-able in kibana or just have the table in the dom?

File: packages/charts/src/components/accessibility/cartesian_data_table.tsx [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+              </tr>
+            );
+          })}
+        </tbody>
+        <tfoot>
+          <tr>
+            <td colSpan={totalColumns}>
+              <button type="submit" onClick={() => handleMoreData()} tabIndex={-1}>
+                Click to show more data
+              </button>
```
Comment (nickofthyme) [inline packages/charts/src/components/accessibility/cartesian_data_table.tsx:106]: Is this necessary? Why not just show all the data with a scroll overflow?

**Review by nickofthyme — COMMENTED** [review]
Still reviewing just leaving early comments.

File: packages/charts/src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx (hunk 1 above)
Comment (rshen91) [reply packages/charts/src/chart_types/xy_chart/renderer/canvas/xy_chart.tsx:183]: AFAIK it's so there's the visual table for debugging vs having to navigate through the DOM.

File: packages/charts/src/components/accessibility/cartesian_data_table.tsx (hunk 2 above)
Comment (rshen91) [reply packages/charts/src/components/accessibility/cartesian_data_table.tsx:106]: I think it's easier with keyboard navigation to have a button to click through voiceover to get the next set of data but I can explore the scroll overflow idea and report back 👍🏻

Comment (nickofthyme) [reply packages/charts/src/components/accessibility/cartesian_data_table.tsx:106]: But considering the screen reader case where someone is not able to see the chart and the screenreader is used. Are they gonna have to keep pressing the "more" button? If so maybe we limit it to more only over `100` items or something. What do you think?

Comment (nickofthyme) [reply packages/charts/src/components/accessibility/cartesian_data_table.tsx:106]: Discussed on zoom

Comment (rshen91) [reply packages/charts/src/components/accessibility/cartesian_data_table.tsx:106]: Thank you for the zoom! I will make a see previous x value button - that makes a lot of sense thank you!

**Comment by markov00** [comment]
hey @rshen91 is this still a WIP or we can review it?
I see two missing checkbox in the description, do we need to wait for them for the review?

**Comment by rshen91** [comment]
> hey @rshen91 is this still a WIP or we can review it? I see two missing checkbox in the description, do we need to wait for them for the review?

Please feel free to review this thank you!

File: packages/charts/src/components/accessibility/cartesian_data_table.tsx [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+  const totalColumns: number = isSmallMultiple ? (countOfCol += 3) : countOfCol;
+
+  /**
+   * Data table for small data sets (less than 20 data points) with only one series
+   */
+  const smallDataTableSingleSeries = (
+    <div
+      className={`echScreenReaderOnly ${debug ? 'echScreenReaderOnlyDebug' : ''} echScreenReaderTable`}
+      aria-live="polite"
+      tabIndex={-1}
```
Comment (1Copenut) [inline packages/charts/src/components/accessibility/cartesian_data_table.tsx:102]: I'm seeing a few `tabindex="-1"` declarations in the table and would like to learn more about how focus is being managed, and how users will navigate the table and change the data point to previous or next point.

File: packages/charts/src/components/accessibility/cartesian_data_table.tsx [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+                  <>
+                    <th scope="row" colSpan={2}>
+                      {values[count].smPanelTitle}
+                    </th>
+                  </>
+                )}
+                {isSmallMultiple && cartesianData.smallMultipleTitle.length === 1 && (
+                  <th scope="row">{values[count].smPanelTitle}</th>
+                )}
+                <th scope="row">{label}</th>
```
Comment (1Copenut) [inline packages/charts/src/components/accessibility/cartesian_data_table.tsx:260]: Is there a way to conditionally make this a `<th scope="row">` element if the previous `isSmallMultiple` prop is false? Looking at the markup, I could see a situation where we have two TH elements that are scoped to the row. That might cause screen readers to behave unexpectedly.

File: packages/charts/src/components/accessibility/cartesian_data_table.tsx [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+                  label: boolean | React.ReactChild | React.ReactFragment | React.ReactPortal | null | undefined;
+                  data: boolean | React.ReactChild | React.ReactFragment | React.ReactPortal | null | undefined;
+                  smPanelTitle: boolean | React.ReactChild | React.ReactFragment | React.ReactPortal | null | undefined;
+                },
+                i: any,
+              ) => {
+                return (
+                  <tr key={String(`screen-reader-row--${index} ${i}`)} tabIndex={-1}>
+                    {isSmallMultiple && cartesianData.smallMultipleTitle.length === 2 && (
+                      <td colSpan={2}>{val.smPanelTitle}</td>
```
Comment (1Copenut) [inline packages/charts/src/components/accessibility/cartesian_data_table.tsx:194]: Same comment about possibly changing this to `<th scope="row">` for the first column from line#133

File: packages/charts/src/components/accessibility/cartesian_data_table.tsx [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+        </thead>
+        <tbody>
+          {cartesianData.data.map(({ values }, index) => {
+            {
+              return values.map(
+                (val: { xValue: any; formatted: any; raw: any; smPanelTitle: string | undefined }, i: number) => {
+                  return (
+                    <tr key={String(`screen-reader-row--${index} ${i}`)} tabIndex={-1}>
+                      {isSmallMultiple && cartesianData.smallMultipleTitle.length === 2 && (
+                        <td colSpan={2}>{val.smPanelTitle}</td>
```
Comment (1Copenut) [inline packages/charts/src/components/accessibility/cartesian_data_table.tsx:133]: Could we make the first cell a `<th scope="row">` for the small multiple condition? I like tables that scope rows because it adds another layer of richness to the information read back.

**Review by 1Copenut — CHANGES_REQUESTED** [review]
@rshen91 Thank you for the work and care you put into this PR! I've got a few items that are questions as much as change requests. Please feel free to take what's relevant and discard what's not.

Holler at me anytime for a quick chat or another review.

**Comment by zuhairmahd** [comment]
I have tested with JAWS and NVDA.  Both screen readers read the table descriptions, but it is important to note that no actual tables are present in the DOM -- I am told this was intended behaviour because the tables are images.

File: packages/charts/src/components/accessibility/cartesian_data_table.tsx (hunk 3 above)
Comment (rshen91) [reply packages/charts/src/components/accessibility/cartesian_data_table.tsx]: Thank you I think I mixed up the appropriate tab index and fixed in 463fb09. Let me know what you think!

**Comment by rshen91** [comment]
The `GroupBy` component provides the vertical axes titles in small multiples cases such as https://elastic.github.io/elastic-charts/?path=/story/small-multiples-alpha--grid-lines&globals=theme:light vs being available on the axis spec. Opening an issue to track small multiple screen reader data table for the future. https://github.com/elastic/elastic-charts/issues/1557

**Review by 1Copenut — APPROVED** [review]
👍 LGTM! I like the markup changes for the table rows, having that scoped TH for each body row will be a strong addition. Thank you for the quick turnaround and logging additional issues to track questions.

**Comment by rshen91** [comment]
jenkins test this

File: packages/charts/src/components/accessibility/partition_data_table.tsx [hunk 7]
```diff
@@ -90,7 +90,7 @@ const ScreenReaderPartitionTableComponent = ({
             .slice(0, rowLimit)
             .map(({ panelTitle, depth, label, parentName, valueText, percentage }, index) => {
               return (
-                <tr key={`row--${index}`} ref={rowLimit === index ? tableRowRef : undefined} tabIndex={-1}>
+                <tr key={String(`row--${index}`)} ref={rowLimit === index ? tableRowRef : undefined} tabIndex={-1}>
```
Comment (nickofthyme) [inline packages/charts/src/components/accessibility/partition_data_table.tsx:93]: This should not be needed.

```suggestion
                <tr key={`row--${index}`} ref={rowLimit === index ? tableRowRef : undefined} tabIndex={-1}>
```

File: playground/playground.tsx [hunk 8]
```diff
@@ -8,8 +8,59 @@
...
+import { Chart } from '../packages/charts/src/components/chart';
+import { ScaleType } from '../packages/charts/src/scales/constants';
+import { AreaSeries, Settings } from '../packages/charts/src/specs';
+// import { Position } from '../packages/charts/src/utils/common';
+import { KIBANA_METRICS } from '../packages/charts/src/utils/data_samples/test_dataset_kibana';
+import { useBaseTheme } from '../storybook/use_base_theme';
 
 export function Playground() {
-  return <Example />;
+  const baseTheme = useBaseTheme();
```
Comment (nickofthyme) [inline playground/playground.tsx:19]: Could you restore this file? I think we should just `.gitignore` it but that's a task for later.

File: packages/charts/src/chart_types/xy_chart/state/selectors/get_screen_reader_data.ts [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+    return ` the ${position} axis is titled ${title}`;
+  });
+  // axesTitles length is currently misleading for small multiple series
+  return `${
+    isSm
+      ? `This chart is a small multiple series.`
+      : `The chart has ${axesTitles.length} axes ${hasTitles.length} axes have titles${
+          hasTitles.length === 0 ? '.' : ':'
+        }`
+  }${getTitles}`;
```
Comment (nickofthyme) [inline packages/charts/src/chart_types/xy_chart/state/selectors/get_screen_reader_data.ts:237]: Just thought, this looks kinda wonky. It might be easier to read if you pull the first value portion into a variable assignment before.

File: packages/charts/src/components/__snapshots__/chart.test.tsx.snap [hunk 10]
```diff
@@ -68,27 +68,98 @@ exports[`Chart should render the legend name test 1`] = `
                 </Connect(CursorBand)>
                 <Connect(XYChart) forwardCanvasRef={{...}}>
                   <XYChart forwardCanvasRef={{...}} isRTL={false} initialized={true} isChartEmpty={false} debug={true} geometries={{...}} geometriesIndex={{...}} theme={{...}} chartContainerDimensions={{...}} highlightedLegendItem={[undefined]} rotation={0} renderingArea={{...}} chartTransform={{...}} axesSpecs={{...}} perPanelAxisGeoms={{...}} perPanelGridLines={{...}} axesStyles={{...}} annotationDimensions={{...}} annotationSpecs={{...}} panelGeoms={{...}} a11ySettings={{...}} onChartRendered={[Function (anonymous)]}>
-                    <figure aria-labelledby={[undefined]} aria-describedby=\\"chart1--defaultSummary\\">
-                      <canvas dir=\\"ltr\\" className=\\"echCanvasRenderer\\" width={151} height={200} style={{...}} role=\\"presentation\\" />
+                    <figure aria-labelledby={[undefined]} aria-describedby=\\"chart1--defaultSummary\\" aria-live=\\"polite\\">
```
Comment (nickofthyme) [inline packages/charts/src/components/__snapshots__/chart.test.tsx.snap:71]: I don't know if this is the right place but just to create a thread....

I don't see any changes to the story files so I'm not sure where the overlay is coming from on the the changed vrt screenshots. Also peculiar that they are all annotation stories. Do you know what is causing this?

image

In any case I don't think it is a good idea to change all these screenshot, nor that the accessibility feature is on by default for the story. But I'd be great if we created a new `integration/tests/accessibility_stories.test.ts` test file to capture the desired use cases.

File: packages/charts/src/components/accessibility/accessibility.test.tsx [hunk 11]
```diff
@@ -177,4 +179,137 @@ describe('Accessibility', () => {
       );
     });
   });
+
```
Comment (nickofthyme) [inline packages/charts/src/components/accessibility/accessibility.test.tsx:182]: 👍🏼

File: packages/charts/src/components/accessibility/cartesian_data_table.tsx [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+        return formatted.hasOwnProperty(val.xValue)
+          ? formatted[val.xValue].push({
+              label,
+              data: val.formatted ?? val.raw,
+              xValue: val.xValue,
+              smPanelTitle: val.smPanelTitle,
+            })
+          : (formatted[val.xValue] = [
+              { label, data: val.formatted ?? val.raw, xValue: val.xValue, smPanelTitle: val.smPanelTitle },
+            ]);
```
Comment (nickofthyme) [inline packages/charts/src/components/accessibility/cartesian_data_table.tsx:49]: This is funky, why are you returning the result of the `push`? Defined to return the new `length` property.

image

The second half of the ternary is also hard to follow with the nested variable assignment. Can we clean this up too?

Is this expected? If not, I'd say just replace this with an if/else block.

File: packages/charts/src/components/accessibility/cartesian_data_table.tsx [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+  }
+  return {
+    a11ySettings: getA11ySettingsSelector(state),
+    cartesianData: getScreenReaderDataSelector(state),
+    debug: getSettingsSpecSelector(state).debug,
+  };
+};
+
+/** @internal */
+export const ScreenReaderCartesianTable = memo(connect(mapStateToProps)(ScreenReaderCartesianTableComponent));
```
Comment (nickofthyme) [inline packages/charts/src/components/accessibility/cartesian_data_table.tsx:314]: Is this meant to always be enabled? Or is there a flag that we can use to determine when to run this extra logic?

**Review by nickofthyme — CHANGES_REQUESTED** [review]
Looking good. Few questions....

File: packages/charts/src/components/accessibility/partition_data_table.tsx (hunk 7 above)
Comment (rshen91) [reply packages/charts/src/components/accessibility/partition_data_table.tsx:93]: Fixed in 51fd846 thank you!

File: playground/playground.tsx (hunk 8 above)
Comment (rshen91) [reply playground/playground.tsx]: Nice! Yes, fixed in 51fd846

File: packages/charts/src/chart_types/xy_chart/state/selectors/get_screen_reader_data.ts (hunk 9 above)
Comment (rshen91) [reply packages/charts/src/chart_types/xy_chart/state/selectors/get_screen_reader_data.ts]: Smart - refactored in 51fd846

File: packages/charts/src/components/accessibility/cartesian_data_table.tsx (hunk 13 above)
Comment (rshen91) [reply packages/charts/src/components/accessibility/cartesian_data_table.tsx:314]: For now it's always enabled. In testing in Kibana it looks like it's easily avoidable by continuing to tab if screen reader users do not want the information. But I think it makes sense to have it enabled by default, thoughts?

File: packages/charts/src/components/__snapshots__/chart.test.tsx.snap (hunk 10 above)
Comment (rshen91) [reply packages/charts/src/components/__snapshots__/chart.test.tsx.snap:71]: In line 183 in the xy-chart.tsx file, there's the clause for if debug is specified to show the component visually. The debug mode then adds the 'echScreenReaderOnlyDebug' class in the cartesian data table component making it visually seen for debugging. I'll look into changing this to remove this overlay and add the extra stories so it can show in only the accessibility cases vs debug generally. Thanks!

File: packages/charts/src/components/accessibility/cartesian_data_table.tsx (hunk 12 above)
Comment (rshen91) [reply packages/charts/src/components/accessibility/cartesian_data_table.tsx:49]: I tried to replace this with an if/else block but we have a ternary rule that is changing it back. 

For charts that have two series, this function is creating a value as an array if the xValue is not yet in the formatted data type. If the xValue has appeared, then it's added to the array. This makes it smoother in the `smallDataTableMultipleSeries` to show information that is a comparison of the different series on the same xValues vs when there is one series. I'm trying to clean it up a little with some commenting to hopefully clear it up a little.

**Comment by nickofthyme** [comment]
Closing to revisit in the future when accessibility is prioritized.
